optimizer/evaluator/evaluator.go

```diff
@@ -0,0 +1,850 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package evaluator
+
+import (
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression/builtin"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/charset"
+	"github.com/pingcap/tidb/util/types"
+	"golang.org/x/text/transform"
+)
+
+// Eval evaluates an expression to a value.
+func Eval(ctx context.Context, expr ast.ExprNode) (interface{}, error) {
+	e := &Evaluator{ctx: ctx}
+	expr.Accept(e)
+	if e.err != nil {
+		return nil, errors.Trace(e.err)
+	}
+	return expr.GetValue(), nil
+}
+
+// EvalBool evalueates an expression to a boolean value.
+func EvalBool(ctx context.Context, expr ast.ExprNode) (bool, error) {
+	val, err := Eval(ctx, expr)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if val == nil {
+		return false, nil
+	}
+	i, err := types.ToBool(val)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return i != 0, nil
+}
+
+func boolToInt64(v bool) int64 {
+	if v {
+		return int64(1)
+	}
+	return int64(0)
+}
+
+// Evaluator is a ast Visitor that evaluates an expression.
+type Evaluator struct {
```

Comment:
Because I think in a visitor, tracing the error is not needed, just know the visitor encountered an error is enough.